src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java

```diff
@@ -518,7 +518,13 @@ private static CcToolchainProvider getToolchainFromPlatformConstraints(
     ResolvedToolchainProviders providers =
         (ResolvedToolchainProviders)
             ruleContext.getToolchainContext().getResolvedToolchainProviders();
-    return (CcToolchainProvider) providers.getForToolchainType(toolchainType);
+    ToolchainInfo toolchainInfo = providers.getForToolchainType(toolchainType);
+    if (toolchainInfo instanceof CcToolchainProvider) {
+      return (CcToolchainProvider) toolchainInfo;
+    }
+
+    ruleContext.ruleError("The selected C++ toolchain is not a cc_toolchain rule");
+    return CcToolchainProvider.EMPTY_TOOLCHAIN_IS_ERROR;
```

Comment:
This looks good per se, but wouldn't this be a problem with other toolchain types? Are all users of toolchains expected to check that they got the kind of toolchain that they expect?

Reply:
In general, users of toolchains should make this check. This is a special case, in that a dummy toolchain was added to make toolchain resolution happy, before the cc_toolchain was fixed.

In general, Skylark rules will all be using ToolchainInfo, which removes the issue with data casts, but raises the issue that ToolchainInfo is basically just a bag of data, and if toolchains are configured improperly, the wrong bag will arrive. I'd like to come up with a general solution for this, but I'm not sure what that would look like.